src/client/Client.cc

```diff
@@ -6890,26 +6905,42 @@ void Client::fill_dirent(struct dirent *de, const char *name, int type, uint64_t
 
 void Client::_readdir_next_frag(dir_result_t *dirp)
 {
-  frag_t fg = dirp->frag();
+  frag_t fg = dirp->buffer_frag;
 
-  // advance
-  dirp->next_frag();
-  if (dirp->at_end()) {
+  if (fg.is_rightmost()) {
     ldout(cct, 10) << "_readdir_next_frag advance from " << fg << " to END" << dendl;
+    dirp->set_end();
+    return;
+  }
+
+  // advance
+  fg = fg.next();
+  ldout(cct, 10) << "_readdir_next_frag advance from " << dirp->buffer_frag << " to " << fg << dendl;
+
+  if (dirp->hash_order()) {
+    int64_t new_offset = dir_result_t::make_fpos(fg.value(), 2, true);
+    if (dirp->offset < new_offset) // don't decrease offset
+      dirp->offset = new_offset;
```

Comment:
Under what scenario could new_offset be less than dirp->offset? I don't really see how that could happen without us also having potentially skipped entries in the next frag. (Mostly, I don't see how that could happen, since we know we're going in hash order.)